[RPC Framework] Expose a Python API for device map getter

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57288 [RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided
* **#57179 [RPC Framework] Expose a Python API for device map getter**

Expose a Python API to get the device map and unblock RemoteModule work.

See: https://github.com/pytorch/pytorch/pull/56854#issuecomment-827762398

Additionally, add a const decorator for the C++ getter.

#Original PR issue: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28070160](https://our.internmc.facebook.com/intern/diff/D28070160/)

Commit:
[RPC Framework] Expose a Python API for device map getter

Expose a Python API to get the device map and unblock RemoteModule work.

See: https://github.com/pytorch/pytorch/pull/56854#issuecomment-827762398

Additionally, add a const decorator for the C++ getter.

Differential Revision: [D28070160](https://our.internmc.facebook.com/intern/diff/D28070160/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -684,6 +684,12 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
               TensorPipeAgent::getWorkerInfos,
           py::call_guard<py::gil_scoped_release>())
+      .def(
+          "_get_device_map",
```
curious, do you need the device map info, or all you need is `_supports_cuda`?

SciPioneer replied:
Thanks for pointing this out! I don't have much context here.

Is `_supports_cuda` here equivalent to `device_map` not empty?

@lw Any suggestion?

lw replied:
If we want to be precise there are various levels of detail. We could have a single boolean `_supports_cuda` field which says whether the agent understands CUDA tensors. We could have a `_devices` fields which says which devices the agent is allowed to operate on (and, if this list contains a CUDA device, then the agent supports CUDA). Or we could have `_devices_for_remote(remote_worker)` which returns the set of devices that can be used when making an RPC call to a certain worker (which would be the keys of the device map for that worker, and would be a subset of `_devices`). Or we could just expose the full device map.

Any of the option above "works", it just depends what @SciPioneer needs.

SciPioneer replied:
> If we want to be precise there are various levels of detail. We could have a single boolean `_supports_cuda` field which says whether the agent understands CUDA tensors. We could have a `_devices` fields which says which devices the agent is allowed to operate on (and, if this list contains a CUDA device, then the agent supports CUDA). Or we could have `_devices_for_remote(remote_worker)` which returns the set of devices that can be used when making an RPC call to a certain worker (which would be the keys of the device map for that worker, and would be a subset of `_devices`). Or we could just expose the full device map.
> 
> Any of the option above "works", it just depends what @SciPioneer needs.

Thanks for the detailed explanation! My current implementation exposes the full device map of the destination worker to `RemoteModule`, and it checks whether the device map is empty. I verified that this works for me.

Commit:
Update on "[RPC Framework] Expose a Python API for device map getter"


Expose a Python API to get the device map and unblock RemoteModule work.

See: https://github.com/pytorch/pytorch/pull/56854#issuecomment-827762398

Additionally, add a const decorator for the C++ getter.

#Original PR issue: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28070160](https://our.internmc.facebook.com/intern/diff/D28070160/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -684,6 +690,12 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
           py::call_guard<py::gil_scoped_release>())
+      .def(
+          "_get_device_map",
+          (tensorpipe::DeviceMap(TensorPipeAgent::*)(const WorkerInfo& dest)
```
This looks OK to me (as it's a private API). I am still curious to see how this is used in RemoteModule. Since this takes `dst` WorkerInfo as an argument, as it mean `RemoteModule` only cares about the device map for a specific worker (maybe the worker where the module lives in?) and won't need to loop over all workers?

mrshenli replied:
Just saw the PR on top. Look so, it just need the device map for one worker.

mrshenli reviewed (approved):
stamp to unblock
